xds/internal/xdsclient/transport/transport.go

```diff
@@ -426,13 +449,22 @@ func (t *Transport) recv(stream adsStream) bool {
 	for {
 		resources, url, rVersion, nonce, err := t.recvAggregatedDiscoveryServiceResponse(stream)
 		if err != nil {
-			t.adsStreamErrHandler(err)
+			// Note that we do not consider it an error if the ADS stream was closed after
+			// having received a response on the stream. This is because there are legitimate
+			// reasons why the server may need to close the stream during normal operations,
+			// such as needing to rebalance load or the underlying connection hitting its
+			// max connection age limit .
+			// (see [gRFC A9](https://github.com/grpc/proposal/blob/master/A9-server-side-conn-mgt.md)).
+			if msgReceived {
+				err = xdsresource.NewErrorf(xdsresource.ErrTypeStreamFailedAfterRecv, err.Error())
+			}
+			t.onErrorHandler(err)
```

Comment:
We should be calling the error handler only if `msgReceived` is `true`, right? Or am I missing something?

Reply:
I think the authority onErrorHandler callback should be called regardless of whether its the first time or nth time. This is to stop all watch timers and decide whether to propagate the errors based on the err type. If errType == ErrTypeStreamFailedAfterRecv, we return from `onErrorHandler` without calling the watchers.